### ISSUE TITLE
Language extension guidelines

### ISSUE THREAD

**Issue by dbaeumer** [Microsoft/vscode-docs#763]
See #702, comments from Dirk & Martin
- we like the structure (introduction, declarative supports, programmatic supports)
- introduction & declarative supports are good.
- Programmatic Language Support is trying too much at once

Proposing the following new structure: 

- clarify the four different approaches how to implement language smartness. For all approaches list one complete example. The approaches are: in extension host, in a language server process speaking the LSP(best would be a non Node server), in the extension host spawning external processes (like Go), in a language server speaking a different protocol (like tsserver)

- list all possible language features without implementation references (keep the animated  gif)

- write child documents for the four implementation possibilities. In the child documents link to existing extensions or samples demonstrating implementation variants (for example Go extension, json extension, linter, ...)

**Comment by dbaeumer** [comment]
@aeschli FYI.

**Comment by gregvanl** [comment]
Seems related to https://github.com/Microsoft/vscode-docs/issues/92

**Comment by vossad01** [comment]
This does seem like a more detailed implementation request for #92.  I understand this is somewhat complicated given this is really documenting the integration between systems/components so it becomes somewhat a question of who is responsible for the documentation.

Nonetheless, a solid example using crossing languages would have been useful for me in figuring out how to connect the pieces for my own development.  If it can help, I have [forked the `vscode-languageserver-node-example` repository used by the documentation and reimplemented the server component in Java](https://github.com/vossad01/vscode-languageserver-java-example) using LSP4J.  I went a little out of my way design-wise to try to help match the layout of the TypeScript so it would be easier to compare the implementations.  I believe it preserves all functionality, though I just got everything working.  I will probably learn more concerning best practices as I move on to implementing my actual language server.
